internal/yunikorn/rest_test.go

```diff
@@ -126,6 +127,640 @@ func TestRESTClient_GetApplication(t *testing.T) {
+		},
+		{
+			name: "Server Error",
+			setup: func() *httptest.Server {
+				return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					http.Error(w, "server error", http.StatusInternalServerError)
+				}))
+			},
+			expected:       nil,
+			wantErr:        true,
+			expectedErrMsg: "yunicorn api returned non-OK status code: 500",
+		},
+		{
+			name: "Unexpected JSON",
+			setup: func() *httptest.Server {
+				return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					type unexpected struct {
+						Unexpected string `json:"unexpected"`
+					}
+					err := json.NewEncoder(w).Encode(&[]unexpected{{Unexpected: "unexpected"}})
+					if err != nil {
+						http.Error(w, err.Error(), http.StatusInternalServerError)
+					}
+				}))
+			},
+			expected:       nil,
+			wantErr:        true,
+			expectedErrMsg: "json: cannot unmarshal array into Go value of type dao.PartitionQueueDAOInfo",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ts := tt.setup()
+			defer ts.Close()
+
+			client := NewRESTClient(getMockServerYunikornConfig(t, ts.URL))
+
+			queues, err := client.GetPartitionQueues(context.Background(), "testPartition")
+			if tt.wantErr {
+				require.Error(t, err)
+				assert.Equal(t, tt.expectedErrMsg, err.Error())
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, tt.expected, queues)
+			}
+		})
+	}
+}
+
+func TestRESTClient_GetPartitionNodes(t *testing.T) {
+	tests := []struct {
+		name           string
+		setup          func() *httptest.Server
+		expected       []*dao.NodeDAOInfo
+		wantErr        bool
+		expectedErrMsg string
+	}{
+		{
+			name: "200 OK Response",
+			setup: func() *httptest.Server {
+				return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					response := []*dao.NodeDAOInfo{
+						{
+							NodeID: "node1",
+						},
+						{
+							NodeID: "node2",
+						},
+					}
+					err := json.NewEncoder(w).Encode(response)
+					if err != nil {
+						http.Error(w, err.Error(), http.StatusInternalServerError)
+					}
+				}))
+			},
+			expected: []*dao.NodeDAOInfo{
+				{
+					NodeID: "node1",
+				},
+				{
+					NodeID: "node2",
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Server Error",
+			setup: func() *httptest.Server {
+				return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					http.Error(w, "server error", http.StatusInternalServerError)
+				}))
+			},
+			expected:       nil,
+			wantErr:        true,
+			expectedErrMsg: "yunicorn api returned non-OK status code: 500",
+		},
+		{
+			name: "Unexpected JSON",
+			setup: func() *httptest.Server {
+				return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					type unexpected struct {
+						Unexpected string `json:"unexpected"`
+					}
+					err := json.NewEncoder(w).Encode(&unexpected{Unexpected: "unexpected"})
```

Comment:
```suggestion
func write(t *testing.T, response any) {
  err := json.NewEncoder(w).Encode(response)
  if err != nil {
    t.Errorf("error writing response: %v", err)
  }
}
```

An error should never appear here, if it does, we are probably trying to return some malformed or invalid response and we want to fail.

Reply:
Hi, I could not understand this change properly. 
Yes, it will not return any error, because we are encoding a valid JSON. The purpose of this test was to test what happens if we return any JSON structure that is not `[]*dao.PartitionInfo` (for example)
As this is a test server setup, I was wondering if it still matters what error response we write.

Reply:
extracted repeated logic into `writeResponse()`